### PR TITLE
Replaced hard-coded global downtime schedule with cached JSON in S3

### PR DESCRIPTION
**Pull request by U-DON**
## Description
We were previoously hard-coding a downtime window object in a `scheduledDowntimeWindows`. Now, we will fetch the downtime schedule in a JSON file that's [cached by the API](https://github.com/department-of-veterans-affairs/vets-api/blob/62686483c1f401bb2b2bac8e00fe2aa6649851b0/app/workers/pager_duty/cache_global_downtime.rb) with a job that runs [every hour](https://github.com/department-of-veterans-affairs/vets-api/blob/62686483c1f401bb2b2bac8e00fe2aa6649851b0/config/sidekiq_scheduler.yml#L107).

Any ongoing global downtime window can now be retrieved by calling `getCurrentGlobalDowntime` in `src/platform/monitoring/DowntimeNotification/util/helpers.js`.

## Testing done
Unit tests. Verified that JSON file can be fetched.

## Acceptance criteria
- [ ] Global downtime windows should not be hard-coded.

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/gi/containers/GiBillApp.jsx [hunk 1]
```diff
@@ -105,10 +103,7 @@ export class GiBillApp extends React.Component {
               facilityCode={facilityCode}
               location={this.props.location}
             />
-            <DowntimeNotification
-              appTitle={'GI Bill Comparison Tool'}
-              dependencies={[externalServices.global]}
-            >
+            <DowntimeNotification appTitle={'GI Bill Comparison Tool'}>
```
Comment (U-DON) [inline src/applications/gi/containers/GiBillApp.jsx]: I'm guessing we want to account for global downtime with GIBCT, which is why I left the DTN here without any dependencies.

But why isn't there a dependency for it as an individual app? Is that dependency GIDS, and is there a PagerDuty service that corresponds to it?

File: src/platform/monitoring/DowntimeNotification/actions/index.js [hunk 2]
```diff
@@ -54,47 +51,24 @@ export function dismissDowntimeWarning(appTitle) {
...
-      dispatch({
-        type: ERROR_SCHEDULE_DOWNTIME,
-        data: globalDowntimeData,
-      });
+      const response = await apiRequest('/maintenance_windows');
+      data = response.data;
+    } catch (error) {
+      // Probably in a test environment and the route isn't mocked.
+    } finally {
+      dispatch({ type: RECEIVE_SCHEDULED_DOWNTIME, data });
```
Comment (U-DON) [inline src/platform/monitoring/DowntimeNotification/actions/index.js]: This is basically a revert to before we tried to incorporate the hard-coded maintenance window into this action creator.

File: src/platform/monitoring/DowntimeNotification/util/helpers.js [hunk 3]
```diff
@@ -104,19 +107,36 @@ export function getSoonestDowntime(serviceMap, serviceNames) {
...
-) => {
-  const { downtimeStart, downtimeEnd } = downtimeWindow;
-
-  return (
-    !environment.isLocalhost() &&
-    moment().isAfter(downtimeStart) &&
-    moment().isBefore(downtimeEnd)
-  );
-};
+export const getCurrentGlobalDowntime = (() => {
```
Comment (U-DON) [inline src/platform/monitoring/DowntimeNotification/util/helpers.js]: @ncksllvn @kelsonic Just a heads up that we've set up a way to configure and retrieve global downtime windows without hard-coding. The way this works is that we can set maintenance windows in PagerDuty for the "Internal: API" service, which is represented by the `global` service name, and those maintenance windows will be stored in a file in S3 as JSON.

File: src/platform/monitoring/DowntimeNotification/util/helpers.js [hunk 4]
```diff
@@ -104,19 +107,36 @@ export function getSoonestDowntime(serviceMap, serviceNames) {
...
+  const MAINTENANCE_WINDOWS_SUBDOMAINS = Object.freeze({
+    [ENVIRONMENTS.VAGOVDEV]: 'dev-va-gov-maintenance-windows',
+    [ENVIRONMENTS.VAGOVSTAGING]: 'staging-va-gov-maintenance-windows',
+    [ENVIRONMENTS.VAGOVPROD]: 'prod-va-gov-maintenance-windows',
+  });
+
+  const subdomain = MAINTENANCE_WINDOWS_SUBDOMAINS[environment.BUILDTYPE];
+
+  const maintenanceWindowsUrl = subdomain
+    ? `https://${subdomain}.${BUCKET_BASE_URL}/maintenance_windows.json`
```
Comment (cvalarida) [inline src/platform/monitoring/DowntimeNotification/util/helpers.js]: How does this file get it here?

File: src/platform/monitoring/DowntimeNotification/util/helpers.js [hunk 5]
```diff
@@ -1,8 +1,11 @@
+import camelCaseKeysRecursive from 'camelcase-keys-recursive';
```
Comment (cvalarida) [inline src/platform/monitoring/DowntimeNotification/util/helpers.js]: Oh. Didn't know that was a thing. :cool:

File: src/platform/monitoring/DowntimeNotification/util/helpers.js [hunk 6]
```diff
@@ -104,19 +107,36 @@ export function getSoonestDowntime(serviceMap, serviceNames) {
...
+    [ENVIRONMENTS.VAGOVPROD]: 'prod-va-gov-maintenance-windows',
+  });
+
+  const subdomain = MAINTENANCE_WINDOWS_SUBDOMAINS[environment.BUILDTYPE];
+
+  const maintenanceWindowsUrl = subdomain
+    ? `https://${subdomain}.${BUCKET_BASE_URL}/maintenance_windows.json`
+    : null;
+
+  const includesCurrentTime = ({ startTime, endTime }) =>
```
Comment (cvalarida) [inline src/platform/monitoring/DowntimeNotification/util/helpers.js]: Might be worth calling out that this assumes the `maintenance_windows.json` is formatted a specific way.
```json
[
  {
    "startTime": "timestamp",
    "endTime": "timestamp"
  }
]
```

File: src/platform/monitoring/DowntimeNotification/util/helpers.js [hunk 7]
```diff
@@ -104,19 +107,36 @@ export function getSoonestDowntime(serviceMap, serviceNames) {
...
+    moment().isAfter(startTime) && moment().isBefore(endTime);
+
+  return async () => {
```
Comment (cvalarida) [inline src/platform/monitoring/DowntimeNotification/util/helpers.js]: Why return a closure from an immediately-invoked (with no parameters) function?

File: src/platform/monitoring/DowntimeNotification/util/helpers.js [hunk 8]
```diff
@@ -104,19 +107,36 @@ export function getSoonestDowntime(serviceMap, serviceNames) {
...
+    try {
+      const response = await fetch(maintenanceWindowsUrl);
+      const data = camelCaseKeysRecursive(await response.json());
+      return data.find(includesCurrentTime) || null;
+    } catch (error) {
+      return null;
```
Comment (cvalarida) [inline src/platform/monitoring/DowntimeNotification/util/helpers.js]: Should we also send an event to Sentry?

**Review by cvalarida — COMMENTED** [review]
This looks great!

My main question here: "How can we tell if it's pulling the `maintenance_windows.json` from all three environments without integration tests?" The unit tests are great, but how can we tell it's fetching as expected?

File: src/applications/gi/containers/GiBillApp.jsx (hunk 1 above)
Comment (U-DON) [reply src/applications/gi/containers/GiBillApp.jsx]: There is a "Component: GI Bill Comparison Tool (GIBCT|GIDS)" service in PagerDuty that doesn't get propagated down from the API as one of the services that have statuses.

**Comment by U-DON** [comment]
For any VSA devs reviewing changes in the apps, the changes are mostly just removing the `global` service as an explicit downtime dependency. Checking for the status of that service will just be default behavior in the `DowntimeNotifications` component.

**Review by mojaray2k — APPROVED** [review]

File: src/applications/gi/containers/GiBillApp.jsx (hunk 1 above)
Comment (rianfowler) [reply src/applications/gi/containers/GiBillApp.jsx]: i think we just document how to set up DTN and let the education team know about it. AFAIK we wan't tracking this yet.

**Review by rianfowler — APPROVED** [review]

**Pull request merged**